Issue/11559 product sales last 30 days

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->
Part II of : #11559 
Part I was implemented and merged here: https://github.com/woocommerce/woocommerce-android/pull/11582
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
In part I https://github.com/woocommerce/woocommerce-android/pull/11582 we added a new card to display the product stock. In this PR I'm adding the missing logic to display the product sales for the last 30 days. 

![Screenshot 2024-05-30 at 23 29 48](https://github.com/woocommerce/woocommerce-android/assets/2663464/aa7afe0f-73f7-4c88-9166-d5687360c15e)

In order to display product sales we need to make

2 Missing parts in this PR that I'll tackle in subsequent PRs: 
- Disabling `Stock` card if the store doesn't have any products or orders yet. Currently, if the store doesn't have any product or orders in place and the stock card is added to the dashboard the `error` state is displayed. So this is a good scenario for testing the error state of the card. 
- `ProductStockRepository` unit tests 
- If the product is a variation, upon clicking on it, navigate directly to the variation instead of the parent product. I made a quick attempt to implement this by directly passing the `variationId` to the product detail navigation event, but didn't work as expected. A lot of information was missing when the variation was opened. Will look into this further. 

### Testing instructions
<!-- Step-by-step testing instructions. When necessary, break out individual scenarios that need testing, and consider including a checklist for the reviewer to go through. -->

1. Log into a site that has products, product variations, and orders placed in the last 30 days. 
2. Edit the stock of the different product and set some to have low stock (1 or 2 units left), some out of stock, and some in back order. 
3. Add `Stock` card to the dashboard
4. Change between the different stock statuses from the card filter and verify the data information is correct.

JorgeMucientes committed:
Fetch sales and product thumbnail for productIds

JorgeMucientes committed:
Fetch variation sales when product is a variation

JorgeMucientes committed:
Get the correct sales value for products and variations

JorgeMucientes committed:
Update gradle changeset

JorgeMucientes committed:
Handle error when fetching variation sales

JorgeMucientes committed:
Remove magic number

JorgeMucientes committed:
Update UI to show sales for last 30 days in dashboard stock card

JorgeMucientes committed:
Minor UI updates on stock card

JorgeMucientes committed:
Handle error actions when fetching stock

JorgeMucientes committed:
Minor padding adjustments

JorgeMucientes committed:
Minor padding adjustments

JorgeMucientes committed:
Handle product stock items taps

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/stock/ProductStockRepository.kt:
```diff
@@ -1,53 +1,141 @@
-    val imageUrl: String?,
-    val itemsSold: Int?
-) : Parcelable
+    private suspend fun getProductSalesReports(stockReport: ProductStockItems):
```
Context on why and how this works. 
- When fetching `wc-analytics/reports/stock` if a product has a `parentId` != 0 then the product is a variation. 
- For a variation, to get the number of items sold, we should request wc-analytics/reports/variations
- For non variable products we fetch the sales stats from `/wc-analytics/reports/products` 

In the current implementation we check what type of products the `stock` report has and fire the needed requests in parallel. 

Known [API bug](https://github.com/woocommerce/woocommerce/issues/47843): If a product has more than 3 variation attributes, then the variation products name returned from the API will be incomplete and won't have appended the variation attributes. We've reported a but here. Hope it gets fixed soon.

JorgeMucientes committed:
Rename field

JorgeMucientes committed:
Make objects, data objects

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/stock/DashboardProductStockCard.kt:
```diff
@@ -44,24 +54,53 @@ fun DashboardProductStockCard(
     onProductClicked: (ProductStockItem) -> Unit,
-    modifier: Modifier = Modifier
+    modifier: Modifier = Modifier,
+    onContactSupportClicked: () -> Unit,
+    onRetryClicked: () -> Unit,
```
np, I think we should move these above the `modifier` argument, as `modifier` is optional, and these are not supposed to be trailing lambdas IMO.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/stock/DashboardProductStockCard.kt:
```diff
@@ -182,14 +221,15 @@ fun ProductStockRow(
         )
-        Column(modifier = Modifier.padding(start = 8.dp)) {
-            Row(modifier = Modifier.padding(bottom = 4.dp, end = 16.dp)) {
+        Column(modifier = Modifier.padding(start = 8.dp, top = 8.dp)) {
```
a very minor np: having non-symmetric paddings makes the logic harder, IMO. To remove the use of non-symmetric paddings, I would use a `Spacer` with a size of `8.dp` instead of the start padding.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/stock/ProductStockRepository.kt:
```diff
@@ -1,53 +1,141 @@
+                getProductSales(
+                    stockReport
+                        .filter { it.parentId == 0L }
+                        .map { it.productId!! },
```
np, I think using `mapNotNull { it.productId }` would be safer.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/stock/ProductStockRepository.kt:
```diff
@@ -1,53 +1,141 @@
+                getVariationsSales(
+                    stockReport
+                        .filter { it.parentId != 0L }
+                        .map { it.productId!! },
```
Same as above.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/stock/ProductStockRepository.kt:
```diff
@@ -1,53 +1,141 @@
+                )
+            }
+            val variationSalesDeferred = async {
+                getVariationsSales(
+                    stockReport
+                        .filter { it.parentId != 0L }
+                        .map { it.productId!! },
+                    startDate,
+                    endDate
+                )
```
This logic means we'll perform fetching for variation reports even when we don't have any variations, I think we should first check if the filtered list is not empty before making the API call, WDYT?

We can also apply the same logic for the products reports as well 🤔

hichamboushaba replied:
So after testing the feature, it's not just about the optimization of not doing an unneeded API call, the API call fails when we don't have any variations, the API returns the following error: `Invalid parameter(s): per_page` as `per_page` will be `0`

JorgeMucientes replied:
Great catch, I actually noticed that while adding unit tests to the repository and already fixed it here: https://github.com/woocommerce/woocommerce-android/pull/11656

> We can also apply the same logic for the products reports as well

The products report fetching already ensures not making any request if there aren't any products. The check is done inside the function `getProductSales()`. I was planning on adding the same check for `getVariationsSales()`, but I missed it.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/stock/ProductStockRepository.kt:
```diff
@@ -1,53 +1,141 @@
+            if (stockReportResult.isError) {
+                Result.failure(Exception(stockReportResult.error.message))
+            } else {
+                val (productSalesResult, variationSalesResult) = getProductSalesReports(stockReportResult.model!!)
```
I think there is a point missing in the logic: which is checking for the local cache before fetching the sales reports, this was suggested and discussed here p1715852940439619/1715228241.268569-slack-C03L1NF1EA3

The idea is simple: just keep a local copy of the responses (stock and sales reports), this could be a simple `Map<ProductStockStatus, List<ProductStockItem>>`, then after fetching the stock report, just compare the products and their stock quantities with the cache, and if they didn't change, proceed with returning the cached value, otherwise invalidate cache and fetch new reports.

This can be implemented on a separate PR if you prefer.

JorgeMucientes replied:
Good point! Missed that part. I'll address that in a different PR. Thanks

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/stock/DashboardProductStockCard.kt:
```diff
@@ -93,8 +132,8 @@ private fun DashboardProductStockCard(
 
             is DashboardProductStockViewModel.ViewState.Error -> {
                 WidgetError(
```
Just a reminder, we'll need to handle the case of when WCAnalytics are missing, I kept a task for this in the ticket https://github.com/woocommerce/woocommerce-android/issues/11544

hichamboushaba replied:
I can take care of this if you are OK with this 🙂

hichamboushaba replied:
I also have some free time, so I can take care of implementing this change if needed.

JorgeMucientes replied:
No worries I can handle it in the same PR where I handle the in-memory cache to avoid unneded requests.

hichamboushaba reviewed (changes requested):
Nice work @JorgeMucientes, the logic looks good, but there is a blocking [point](https://github.com/woocommerce/woocommerce-android/pull/11631#discussion_r1625791577) that needs to be addressed before I can properly test the changes, and I left some other remarks for you to check (but there are not blockers for this PR).

JorgeMucientes committed:
Apply minor code improvements over UI code

JorgeMucientes committed:
Make mapping calls null safe

JorgeMucientes commented:
Hey @hichamboushaba, thanks for the review and the excellent feedback, as always. I've addressed all the suggestions. About the blocking issue as mentioned above I fixed it in the PR were I added the tests, so no need to add here. 

The rest of the bigger changes (memory cache and handling `WCAnalytics are missing` case) I'll tackle that in a different PR.

hichamboushaba reviewed (approved):
Nice work @JorgeMucientes, let's merge this, and handle the missing parts on other PRs.

One missing part is an empty state, the card looks like this when we don't have any products matching the selected filter:
<img width="331" alt="Screenshot 2024-06-04 at 16 07 02" src="https://github.com/woocommerce/woocommerce-android/assets/1657201/26926207-f304-4eda-9949-7ebadc5af490">

> Disabling Stock card if the store doesn't have any products or orders yet. Currently, if the store doesn't have any product or orders in place and the stock card is added to the dashboard the error state is displayed. So this is a good scenario for testing the error state of the card.

Disabling the card when the site doesn't have products is a good idea, but why would we need to disable the card when the store doesn't have orders? In my experience, this didn't cause any errors, the card worked fine.

JorgeMucientes commented:
@hichamboushaba 

> One missing part is an empty state; the card looks like this when we don't have any products matching the selected filter:

Good catch! I'll make sure to add one

> Disabling the card when the site doesn't have products is a good idea, but why would we need to disable the card when the store doesn't have orders? In my experience, this didn't cause any errors, the card worked fine. 

I don't have a strong opinion about this, but I felt that if the user doesn't have any orders, a card that displays stock and sales isn't very useful to them. But I can disable it only if there are not products 👍🏼

JorgeMucientes committed:
Update FluxC changeset

JorgeMucientes committed:
Use a single product reports store

JorgeMucientes committed:
Update FluxC changeset

JorgeMucientes committed:
Merge branch 'refs/heads/trunk' into issue/11559-product-sales-last-30-days

# Conflicts:
#	build.gradle
